set -Werror for builds

dagitses opened the pull request:
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/pytorch/pytorch/pull/89865).
* __->__ #89865
* #89852
* #89851

set -Werror for builds

Test Plan: Rely on CI.



cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @EikanWang

malfet reviewed (changes requested):
IMO `WERROR` should not be set to true by default in CMakeLists, but rather in CI, which is already the case if one to look at https://github.com/pytorch/pytorch/blob/6d794f6a4ac6c8c36a4521434bf2842090c87789/.jenkins/pytorch/build.sh#L213

dagitses commented:
> IMO `WERROR` should not be set to true by default in CMakeLists, but rather in CI, which is already the case if one to look at
> 
> https://github.com/pytorch/pytorch/blob/6d794f6a4ac6c8c36a4521434bf2842090c87789/.jenkins/pytorch/build.sh#L213

correct it does, which is why this thankfully didn't require *too* many additional changes.

So why do you not think it should be on by default? Wouldn't users generally want to find errors as soon as possible, rather than wait until they push to github and CI runs?

malfet commented:
> So why do you not think it should be on by default? Wouldn't users generally want to find errors as soon as possible, rather than wait until they push to github and CI runs?
Because PyTorch would become unbuildable with new compilers (i.e. every time XCode is updated on my personal mac it uncovers 3-5 new errors that were not detected by the previous iteration of the compiler). So we should advice devs to build with WERROR=1, but it shouldn't be forced upon them, as those newly discovered warnings would not regress existing functionality, would they?

Commit:
set -Winconsistent-missing-override for builds

Summary: This has triggered internally on some PyTorch code.

Test Plan: Rely on CI.

Commit:
set -Werror for builds

Test Plan: Rely on CI.

Commit:
set -Wsuggest-override for builds

Summary: This was flagged by a Meta internal build.

Test Plan: Rely on CI.
